atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractPersistentStreamStore.java

```diff
@@ -200,12 +206,24 @@ protected void storeBlockWithNonNullTransaction(@Nullable Transaction tx, final
         if (tx != null) {
             storeBlock(tx, id, blockNumber, bytesToStore);
         } else {
-            Preconditions.checkNotNull(txnMgr, "Transaction manager must not be null");
-            txnMgr.runTaskThrowOnConflict(
-                    (TransactionTask<Void, RuntimeException>) t1 -> {
-                        storeBlock(t1, id, blockNumber, bytesToStore);
-                        return null;
-                    });
+            attemptToStoreBlockWithNewTransaction(id, blockNumber, bytesToStore);
+        }
+    }
+
+    private void attemptToStoreBlockWithNewTransaction(final long id, final long blockNumber,
+            final byte[] bytesToStore) {
+        Preconditions.checkNotNull(txnMgr, "Transaction manager must not be null");
+
+        for (int attempt = 1; attempt <= NUM_RETRIES_ON_QOS_EXCEPTION; attempt++) {
+            try {
+                txnMgr.runTaskThrowOnConflict(
+                        (TransactionTask<Void, RuntimeException>) t1 -> {
+                            storeBlock(t1, id, blockNumber, bytesToStore);
+                            return null;
+                        });
+            } catch (RateLimitExceededException ex) {
+                Uninterruptibles.sleepUninterruptibly(attempt, TimeUnit.SECONDS);
```

Comment:
Would be good to add some logging and maybe a maximum sleep duration if we make the number configurable.